src/metrics/include/metrics/v2/event.hpp

```diff
@@ -247,70 +285,269 @@ class Collector {
+class MetricsBuffer {
+public:
+    explicit MetricsBuffer(size_t size, const std::string& name)
+        : size{size},
+          name{name},
+          _loading(new std::vector<MetricsArgs<ClockSource>>()),
+          _draining(new std::vector<MetricsArgs<ClockSource>>()) {
+        _draining->reserve(size);
+        _loading->reserve(size);
+    }
+
+    // Thread-safe.
+    void addAt(const typename ClockSource::time_point& finish,
+               OperationEventT<ClockSource> event,
+               size_t workerCount) {
+        _loading_mutex.lock();
+        _loading->emplace_back(finish, std::move(event), workerCount);
+        _loading_mutex.unlock();
+    }
+
+    // Not thread-safe.
+    std::optional<MetricsArgs<ClockSource>> pop(bool force) {
+        refresh(force);
+        if (_draining->empty()) {
+            return std::nullopt;
+        } else {
+            auto ret = std::move(_draining->back());
+            _draining->pop_back();
+            return ret;
+        }
+    }
+
+    const std::string name;
+    const size_t size;
+
+private:
+    void refresh(bool force) {
+        if (_draining->empty()) {
+            _loading_mutex.lock();
+            if (force || _loading->size() >= size * SWAP_BUFFER_PERCENT) {
+                _draining.swap(_loading);
+            }
+            _loading_mutex.unlock();
+        }
+
+        // Maybe a bit nuclear, but this draws a box around the entire grpc system
+        // and errors if it ever backs up enough to slow down an actor thread.
+        if (_draining->size() > size) {
+            std::ostringstream os;
+            os << "Metrics buffer for operation name " << name << " exceeded pre-allocated space"
+               << ". Expected size: " << size << ". Actual size: " << _draining->size()
+               << ". This may affect recorded performance.";
+
+            BOOST_THROW_EXCEPTION(MetricsError(os.str()));
+        }
+    }
+
+    std::unique_ptr<std::vector<MetricsArgs<ClockSource>>> _loading;
+    std::unique_ptr<std::vector<MetricsArgs<ClockSource>>> _draining;
+    std::mutex _loading_mutex;
+};
 /**
  * Primary point of interaction between v2 poplar internals and the metrics system.
  */
 template <typename ClockSource, typename StreamInterface>
 class EventStream {
     using duration = typename ClockSource::duration;
     using OptionalPhaseNumber = std::optional<genny::PhaseNumber>;
+    using ClientPtr = std::shared_ptr<GrpcClient<ClockSource, StreamInterface>>;
 
 public:
     explicit EventStream(const ActorId& actorId,
                          const std::string& name,
                          const OptionalPhaseNumber& phase,
-                         const boost::filesystem::path& pathPrefix)
-        : _name{name}, _stream{name, actorId}, _phase{phase}, _last_finish{ClockSource::now()} {
+                         const boost::filesystem::path& pathPrefix,
+                         ClientPtr grpcClient = ClientPtr(nullptr))
```

Comment:
shouldn't this always be provided?
Another case to consider a ref as well.

Reply:
Ended up making it a ref to a unique_ptr.